pkgs/drain/drain_test.go

```diff
@@ -0,0 +1,215 @@
+		Should(BeFalse(), "Test pod still exists after deletion")
+}
+
+func testPodIsRunning() bool {
+	pod := testutils.GetPod(k8sClient, podName, podNamespace)
+	if pod != nil {
+		return pod.Status.Phase != corev1.PodRunning
+	}
+	return false
+}
+
+func isNodeReady(n corev1.Node) bool {
+	for _, condition := range n.Status.Conditions {
+		if condition.Type == corev1.NodeReady && condition.Status == corev1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func isNodeSchedulable(n corev1.Node) bool {
+	if n.Spec.Unschedulable {
+		return false
+	}
+	for _, taint := range n.Spec.Taints {
+		if taint.Effect == corev1.TaintEffectNoSchedule {
+			return false
+		}
+	}
+	return true
+}
+
+func getNode(c client.Client) corev1.Node {
+	nodes := &corev1.NodeList{}
+	err := c.List(context.Background(), nodes)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(nodes.Items).NotTo(BeEmpty(), "No nodes found in the cluster")
+	return nodes.Items[0]
+}
+
+func TestKindCluster(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Drain helper suite")
+}
+
+var _ = Describe("Drain Interface", Ordered, func() {
+	BeforeAll(func() {
+		var err error
+
+		opts := zap.Options{
+			Development: true,
+		}
+		ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+		// Create a KindCluster
+		testCluster = testutils.KindCluster{Name: "dpu-operator-test-cluster"}
+		restConfig = testCluster.EnsureExists()
+
+		k8sClient, err = client.New(restConfig, client.Options{})
+		Expect(err).NotTo(HaveOccurred())
+		node = getNode(k8sClient)
+	})
+	AfterAll(func() {
+		if os.Getenv("FAST_TEST") == "false" {
+			testCluster.EnsureDeleted()
+		} else {
+			ensureTestPodDeleted(k8sClient)
+		}
+	})
+	Context("When node drain has not been requested", func() {
+		It("should have all nodes available", func() {
+			By("checking that all nodes are ready and schedulable")
+			ctrl.Log.Info("Checking node is ready", "nodeName", node.Name)
+
+			Expect(isNodeReady(node)).To(BeTrue(), "Node is not ready: "+node.Name)
+			Expect(isNodeSchedulable(node)).To(BeTrue(), "Node is not schedulable: "+node.Name)
+		})
+		It("should be able to run a workload", func() {
+			By("scheduling a test workload")
+			ctrl.Log.Info("Creating test pod on node", "nodeName", node.Name)
+			ensureTestPodCreated(k8sClient)
+		})
+	})
+	Context("When node drain and cordon has been requested", func() {
+		It("should drain the node and mark it unschedulable", func() {
+			By("checking that all nodes are not ready and draining")
+			Eventually(func() bool {
+				return testPodIsRunning()
+			}, testutils.TestAPITimeout, testutils.TestRetryInterval).Should(BeTrue(), "Pod did not become running in expected time")
+
+			ctrl.Log.Info("Draining node", "nodeName", node.Name)
+			var err error
+			testDrainer, err = createNewDrainer(restConfig)
+			Expect(err).NotTo(HaveOccurred())
+
+			drainSuccess, err := testDrainer.DrainNode(context.TODO(), &node, true)
```

Comment:
lines 172 to 181, could we move it above 171 so that the context in our BDD framework here makes sense?

Reply:
Done